---
title: "RNA Editing"
author: "Meg Mindlin"
date: "`r Sys.Date()`"
output: pdf_document
---

```{r}
library(ape)
```



```{r}
high.priority=read.csv("high_priority_blasted.csv")
edits.sig=read.csv("sig_edits.csv")
```


```{r}
zf=
edits.sig[
  edits.sig$genomic_aa!=edits.sig$edited_aa&
  edits.sig$gdna_con=="A"&
  edits.sig$edited=="G"&
  grepl("z.*n.*finger",edits.sig$protein_name,ignore.case = T)
,]
```


# making the filtered fasta file
making a fasta file with only the ORFs containing high priority edits
First, I unzip the zip file that has the large fasta file in it
```{bash, eval=F}
unzip rubescens_transcriptome_ORF_ignore_nested.zip
```

then read it into R and make a new fasta that only has the high priority edits
```{r, eval=F}
ORF=read.dna('rubescens_transcriptome_ORF_ignore_nested.fasta', format='fasta')
zf.orf=ORF[gsub("(.*) ORF.*","\\1",labels(ORF)) %in% zf$orf]
zf.aa=trans(zf.orf)
write.dna(zf.orf,"zf_orf.fasta",format="fasta",colsep="")

write.dna(zf.aa,"zf_aa.fasta",format="fasta",colsep="")


```

Then I remove the large fasta.
```{bash, eval=F}
rm rubescens_transcriptome_ORF_ignore_nested.fasta
```

This calls the script that takes the fasta with only the ORF sequences from nonsynonymous A to G significant edits, and makes two fastas one with the edited version (AtoG_orf_edited.fasta), and the unedited version (AtoG_orf_unedited.fasta).
```{bash}
python3 make_edits_in_fasta.py AtoG_nonsyn_sig_edits.csv AtoG_orf.fasta
```

reading in edited and unedited fasta files
```{r}
edited=read.dna('AtoG_orf_edited.fasta',format='fasta')
unedited=read.dna('AtoG_orf_unedited.fasta',format='fasta')
```


```{r}
write.dna(trans(edited),"AtoG_aa_edited.fasta",format="fasta",colsep="")
write.dna(trans(unedited),"AtoG_aa_unedited.fasta",format="fasta",colsep="")
```

```{bash}
cp AtoG_aa_edited.fasta pwm_predict/
cp AtoG_aa_unedited.fasta pwm_predict/
```

```{bash}
cd pwm_predict/
./pwm_predict -m SVMp AtoG_aa_edited.fasta
./pwm_predict -m SVMp AtoG_aa_unedited.fasta
```

```{bash}
cd pwm_predict/
cp AtoG_aa_edited.pwm ..
cp AtoG_aa_unedited.pwm ..
```

```{bash}
python3 scripts/pwm_to_fastq.py AtoG_aa_unedited.pwm AtoG_aa_unedited_zf_targets.fastq
python3 scripts/pwm_to_fastq.py AtoG_aa_edited.pwm AtoG_aa_edited_zf_targets.fastq
```

replacing bases with less than a 50% probability with N
```{bash}
seqtk seq -q 3 -n N AtoG_aa_edited_zf_targets.fastq > AtoG_aa_edited_zf_targets_N.fastq
seqtk seq -q 3 -n N AtoG_aa_unedited_zf_targets.fastq > AtoG_aa_unedited_zf_targets_N.fastq
```

Finding ORFs for which the editing and unedited have different targets
```{bash}
python3 compare_fastq.py
```

```{r}
nonident=read.table("non_identical_headers.txt")[,1]
write.dna(ORF[gsub("(.*) ORF.*","\\1",labels(ORF)) %in% nonident],"Zinc_fingers_that_change.fasta",format="fasta",colsep="")
write.csv(AtoG.nonsyn[AtoG.nonsyn$orf %in% nonident,],"zinc_fingers_that_change.csv",row.names = F)
```

```{r}
edits.sig[edits.sig$orf=="lcl|TRINITY_DN90_c0_g1_i16:425-2539",]
```

```{r}
AtoG.nonsyn[grepl("425-2539",AtoG.nonsyn$orf),]
```


