---
title: "RNA Editing"
author: "Meg Mindlin"
date: "`r Sys.Date()`"
output: html_document
---

```{r}
library(ape)
library(Rsamtools)
library(dplyr)
library(stringr)
library(Biostrings) # For handling FASTQ files
library(ShortRead)
library(perm)
```



```{r}
high.priority=read.csv("high_priority_blasted.csv")
edits.sig=read.csv("sig_edits.csv")
```

# Pulling ORFs that are annotated to be zinc fingers from the SWISSPROT annotations.  However, we don't use this anymore (because we predict them with PWM Predict), so we can probably delete this. But double check first.
```{r}
zf=
edits.sig[
  edits.sig$genomic_aa!=edits.sig$edited_aa&
  edits.sig$gdna_con=="A"&
  edits.sig$edited=="G"&
  grepl("z.*n.*finger",edits.sig$protein_name,ignore.case = T)
,]
```


# making the filtered fasta file
making a fasta file with only the ORFs containing high priority edits
First, I unzip the zip file that has the large fasta file in it
```{bash, eval=F}
unzip rubescens_transcriptome_ORF_ignore_nested.zip
```

then read it into R and make a new fasta that only has the high priority edits
```{r, eval=F}
ORF=read.dna('swissprotORF.fasta', format='fasta') # originally was rubescens_transcriptome_ORF_ignore_nested.fasta
zf.orf=ORF[gsub("(.*) ORF.*","\\1",labels(ORF)) %in% zf$orf]
zf.aa=trans(zf.orf)
write.dna(zf.orf,"zf_orf.fasta",format="fasta",colsep="")

write.dna(zf.aa,"zf_aa.fasta",format="fasta",colsep="")


```

```{r}
write.dna(trans(ORF),"ORF_aa.fasta",format="fasta",colsep="")
```


# copy transcriptome fasta to pwm_predict folder
```{bash}
cp ORF_aa.fasta pwm_predict
```

```{bash}
grep -c "^>" ORF_aa.fasta
```

# how many zinc fingers are in the entire transcriptome
```{bash}
cd pwm_predict/
./pwm_predict -m SVMp ORF_aa.fasta
```

```{bash}
mv pwm_predict/ORF_aa.pwm .
```

```{bash}
grep -c "^>" ORF_aa.pwm 
```

converting the pwm to fastq of the zinc finger targets
```{bash}
python3 scripts/pwm_to_fastq.py ORF_aa.pwm ORF_aa_zf_targets.fastq
```

Finding Zinc Fingers in different sets of edited genes
```{r}


# Read the FASTQ file
fastq_data = readFastq("ORF_aa_zf_targets.fastq")

# Extract the headers from the FASTQ file
fastq_headers = sapply(strsplit(as.character(id(fastq_data)), " "), `[`, 1)

# Getting logical of zinc fingers within edits.sig
edits.sig.zf.logical = unique(edits.sig$orf) %in% fastq_headers

# Count the number of matches
print(paste("Number of genes bearing significant edits:",length(unique(edits.sig$orf))))
print(paste("Number zinc fingers among genes bearing significant edits:",sum(edits.sig.zf.logical)))
sum(edits.sig.zf.logical)/length(unique(edits.sig$orf))


# Getting logical of zinc fingers within edits.sig
high.priority.zf.logical = unique(high.priority$orf) %in% fastq_headers

# Count the number of matches
print(paste("Number of genes bearing high priority (top 50) edits:",length(unique(high.priority$orf))))
print(paste("Number zinc fingers among genes bearing high priority (top 50) edits:",sum(high.priority.zf.logical)))
sum(high.priority.zf.logical)/length(unique(high.priority$orf))


```

```{r}
edits.sig.zf=edits.sig[edits.sig.zf.logical,]

```



Then I remove the large fasta.
```{bash, eval=F}
rm rubescens_transcriptome_ORF_ignore_nested.fasta
```

This calls the script that takes the fasta with only the ORF sequences from nonsynonymous A to G significant edits, and makes two fastas one with the edited version (AtoG_orf_edited.fasta), and the unedited version (AtoG_orf_unedited.fasta).
```{bash}
python3 make_edits_in_fasta.py AtoG_nonsyn_sig_edits.csv AtoG_orf.fasta
```

Reading in edited and unedited fasta files
```{r}
edited=read.dna('AtoG_orf_edited.fasta',format='fasta')
unedited=read.dna('AtoG_orf_unedited.fasta',format='fasta')
```

Then writing out the amino acid translations of those ORFs.
```{r}
write.dna(trans(edited),"AtoG_aa_edited.fasta",format="fasta",colsep="")
write.dna(trans(unedited),"AtoG_aa_unedited.fasta",format="fasta",colsep="")
```

```{bash}
cp AtoG_aa_edited.fasta pwm_predict/
cp AtoG_aa_unedited.fasta pwm_predict/
```

```{bash}
cd pwm_predict/
./pwm_predict -m SVMp AtoG_aa_edited.fasta
./pwm_predict -m SVMp AtoG_aa_unedited.fasta
```

```{bash}
cd pwm_predict/
cp AtoG_aa_edited.pwm ..
cp AtoG_aa_unedited.pwm ..
```

```{bash}
python3 scripts/pwm_to_fastq.py AtoG_aa_unedited.pwm AtoG_aa_unedited_zf_targets.fastq
python3 scripts/pwm_to_fastq.py AtoG_aa_edited.pwm AtoG_aa_edited_zf_targets.fastq
```

replacing bases with less than a 50% probability with N
```{bash}
seqtk seq -q 3 -n N AtoG_aa_edited_zf_targets.fastq > AtoG_aa_edited_zf_targets_N.fastq
seqtk seq -q 3 -n N AtoG_aa_unedited_zf_targets.fastq > AtoG_aa_unedited_zf_targets_N.fastq
```

Finding ORFs for which the edited and unedited have different targets
```{bash}
python3 compare_fastq.py
```

```{r}
nonident=read.csv("non_identical_headers.txt")[,1]
write.dna(ORF[gsub("(.*) ORF.*","\\1",labels(ORF)) %in% nonident],"Zinc_fingers_that_change.fasta",format="fasta",colsep="")
write.csv(AtoG.nonsyn[AtoG.nonsyn$orf %in% nonident,],"zinc_fingers_that_change.csv",row.names = F)
```

Adding @ to the beginning of the headers so that they would they would match
```{bash}
sed -i 's/^/@/' non_identical_headers.txt
```

```{bash}
seqtk subseq AtoG_aa_edited_zf_targets.fastq non_identical_headers.txt > nonident_targets_edited.fastq
seqtk subseq AtoG_aa_unedited_zf_targets.fastq non_identical_headers.txt > nonident_targets_unedited.fastq
```

# Aligning to transcriptome
The next four chunks are done in the Work folder on the genomics computer (/media/work/ruby_trans_alignment)
preparing transcriptome
```{bash}
# Only need to run this once.
cd /media/work/ruby_trans_alignment
STAR --runThreadN 20  	--runMode genomeGenerate  	--genomeDir /media/work/ruby_trans_alignment/genomeDir  	--genomeFastaFiles /media/work/ruby_trans_alignment/Trinity.fasta  	--genomeSAindexNbases 12  --limitGenomeGenerateRAM 204889114890
```


```{bash}
cd /media/work/ruby_trans_alignment
STAR --runThreadN 20 --genomeDir /media/work/ruby_trans_alignment/genomeDir --readFilesIn /media/work/ruby_trans_alignment/nonident_targets_edited.fastq --outFileNamePrefix edited_targets --outSAMtype BAM Unsorted --outSAMprimaryFlag OneBestScore --outSAMmultNmax 1
```

```{bash}
cd /media/work/ruby_trans_alignment
STAR --runThreadN 20 --genomeDir /media/work/ruby_trans_alignment/genomeDir --readFilesIn /media/work/ruby_trans_alignment/nonident_targets_unedited.fastq --outFileNamePrefix unedited_targets --outSAMtype BAM Unsorted --outSAMprimaryFlag OneBestScore --outSAMmultNmax 1
```

copying SAM files back to Genomics_Sync folder
```{bash}
cd /media/work/ruby_trans_alignment
cp edited_targetsAligned.out.bam /home/onthank/Documents/Genomics_Sync/Octopus_rubescens_optic_lobe_RNA_Editing/
cp unedited_targetsAligned.out.bam /home/onthank/Documents/Genomics_Sync/Octopus_rubescens_optic_lobe_RNA_Editing/
```

```{r}
unedited.aligned=scanBam("unedited_targetsAligned.out.bam", param = ScanBamParam(what = c("qname", "rname")))
edited.aligned=scanBam("edited_targetsAligned.out.bam", param = ScanBamParam(what = c("qname", "rname")))
```

```{r}
# Ensure rname columns from BAM alignments are treated as characters
unedited_headers <- unedited.aligned[[1]]$qname  # Headers from unedited BAM
unedited_references <- as.character(unedited.aligned[[1]]$rname)  # References from unedited BAM as characters

edited_headers <- edited.aligned[[1]]$qname  # Headers from edited BAM
edited_references <- as.character(edited.aligned[[1]]$rname)  # References from edited BAM as characters

# Create named vectors for easier lookup
unedited_alignment <- setNames(unedited_references, unedited_headers)
edited_alignment <- setNames(edited_references, edited_headers)

# Extract the truncated portion of nonident headers (before the first space)
truncated_nonident <- sapply(strsplit(nonident, " "), `[`, 1)

# Create the final table, matching on the truncated headers
result_table <- data.frame(
  Sequence = nonident,  # Full headers from nonident
  Edited_Aligned_To = ifelse(truncated_nonident %in% names(edited_alignment),
                             edited_alignment[truncated_nonident], NA),
  Unedited_Aligned_To = ifelse(truncated_nonident %in% names(unedited_alignment),
                               unedited_alignment[truncated_nonident], NA),
  stringsAsFactors = FALSE  # Ensure the resulting table columns are characters
)

# View the result
print(result_table)

# Save the result as a CSV file
write.csv(result_table, "ZF-transcriptome_alignment_table.csv", row.names = FALSE)

```

# ZF transcripts have more significant edits than non-ZF transcripts
We have already made an object called "fastq_headers" that has all of the zinc finger. I need to get each transcript only once.
```{r}
zf.unique=unique(fastq_headers)
```

```{r}
ORF.names=names(ORF)
ORF.names=sub(" .*", "", ORF.names)
not.zf=ORF.names[!ORF.names %in% zf.unique]

notzf.table=table(edits.sig$orf[edits.sig$orf %in% not.zf])
zf.table=table(edits.sig$orf[edits.sig$orf %in% zf.unique])

notzf.zeros=sum((!ORF.names %in% edits.sig$orf)&(!ORF.names %in% zf.unique))
zf.zeros=sum((!ORF.names %in% edits.sig$orf)&(ORF.names %in% zf.unique))
```

```{r}
notzf.numeric=c(as.numeric(notzf.table),rep(0,notzf.zeros))
zf.numeric=c(as.numeric(zf.table),rep(0,zf.zeros))
length(notzf.numeric)+length(zf.numeric)

mean(notzf.numeric)
mean(zf.numeric)

permTS(notzf.numeric,zf.numeric)
```

```{r}

hist(notzf.numeric,breaks=max(notzf.numeric)+1,xlim=c(0,10),freq=FALSE,ylim=c(0,1),col="#a53860")
hist(zf.numeric,breaks=max(zf.numeric)+1,xlim=c(0,10),freq=FALSE,ylim=c(0,1),col="#ffa5ab")

```


```{r}
table(edits.sig$orf)[table(edits.sig$orf)>=10]
```

```{r}
png(filename="edit_distribution_histogram.png",height=1080,width=1920,pointsize=32)
par(mar = c(4, 2, 0, 1),bg=NA) 

# Define breaks
breaks <- seq(0, max(c(notzf.numeric, zf.numeric)) + 1, by = 1)

# Calculate histogram counts
notzf.hist <- hist(notzf.numeric, breaks = breaks, plot = FALSE)
zf.hist <- hist(zf.numeric, breaks = breaks, plot = FALSE)

# Calculate proportions for normalized frequency
notzf.prop <- notzf.hist$counts / sum(notzf.hist$counts)
zf.prop <- zf.hist$counts / sum(zf.hist$counts)

# Plot back-to-back histograms
barplot(
  -notzf.prop, names.arg = notzf.hist$mids, horiz = TRUE, beside = TRUE,
  col = "#a53860", xlim = c(-1, 1), ylim = c(0, 11),
  border = "black", space = 0, xaxt = "n", yaxt="n",
  xlab = "Percent of Transcripts"
)
barplot(
  zf.prop, add = TRUE, horiz = TRUE, beside = TRUE, col = "#da627d",
  border = "black", space = 0, axes=F
)

mtext("Number of acidification-responsive edits",side=2,line=0.5)
for (i in 1:11){
  text(-notzf.prop[i]-0.035,i-0.5,i-1)
}
text(-.50,8,"Non-zinc finger containing transcripts")
text(.50,8,"Zinc finger containing transcripts")
# Add axis and legend
axis(1, at = seq(-1, 1, by = 0.2), labels = abs(seq(-100, 100, by = 20)))
#axis(2,at=seq(from=0.5, to =10.5, by=1), labels=0:10,lwd=0,las=2,line=-1.5)
#legend(
#  "topright", legend = c("Not ZF", "ZF"), fill = c("#a53860", "#ffa5ab"),
#  border = "white"
#)
dev.off()
```



# alignment to genome
```{r}

```



